makefiles/tools/compile_commands.inc.mk

```diff
@@ -0,0 +1,82 @@
+# clangd doesn't automatically include system headers of embedded toolchain. Add this as extra
+# include dir, but allow overriding it
+COMPILE_COMMANDS_INCLUDES_EXTRA ?= -I/usr/$(TARGET_ARCH)/include
+# clangd doesn't support -Wformat-truncation. Just ignore unknown warnings here
+COMPILE_COMMANDS_CFLAGS_FLAGS   ?= -Wno-unknown-warning-option
+# The following list of flags confuse clangd
+COMPILE_COMMANDS_STRIP_FLAGS    ?= -mno-thumb-interwork
+
+# Templates used to generate valid JSON. The contain magic keywords in pointy braces to be replaced
+# by the correct values
+define T_COMPILE_COMMAND
+  {
+    "arguments": [
+      "<COMPILER>",
+ <FLAGS>      "-o",
+      "<OUTPUT>",
+      "-c",
+      "<FILE>"
+    ],
+    "directory": "<DIRECTORY>",
+    "file": "<DIRECTORY>/<FILE>",
+    "output": "<OUTPUT>"
+  }<SEPARATOR>
+
+endef
+
+define T_COMPILE_COMMANDS
+[
+<COMMANDS>]
+endef
+
+define T_FLAGS
+     "<FLAG>",
+
+endef
+
+COMMA := ,
+SRC_TO_OBJ              = $(BINDIR)/$(MODULE)/$(basename $1).o
+
+# List of compiler flags to export
+FLAG_LIST               := $(CFLAGS) $(COMPILE_COMMANDS_CFLAGS_FLAGS) \
+                           $(INCLUDES) $(COMPILE_COMMANDS_INCLUDES_EXTRA)
+ifneq (,$(COMPILE_COMMANDS_STRIP_FLAGS))
+  FLAG_LIST             := $(filter-out $(COMPILE_COMMANDS_STRIP_FLAGS),$(FLAG_LIST))
+endif
+FLAG_LIST               := $(foreach flag,$(FLAG_LIST),$(subst <FLAG>,$(flag),$(T_FLAGS)))
+
+# Functions that insert values into the template. The functions are chained, so that each function
+# inserts one item and calls the next stage. This is more readable than an ultra long expression.
+ADD_COMPILER            = $(subst <COMPILER>,$(CC),$(T_COMPILE_COMMAND))
+ADD_DIRECTORY           = $(subst <DIRECTORY>,$(CURDIR),$(call ADD_COMPILER,$1))
+ADD_OUTPUT              = $(subst <OUTPUT>,$(call SRC_TO_OBJ,$1),$(call ADD_DIRECTORY,$1))
+ADD_SOURCE              = $(subst <FILE>,$1,$(call ADD_OUTPUT,$1))
+ADD_FLAGS               = $(subst <FLAGS>,$(FLAG_LIST),$(call ADD_SOURCE,$1))
+ADD_SEPERATOR           = $(subst <SEPARATOR>,$2,$(call ADD_FLAGS,$1))
+
+SRC_C_ALL               := $(sort $(SRC) $(SRC_NO_LTO))
+SRC_CXX_ALL             := $(sort $(SRCXX))
+
+# This is a bit involved, as the last command must not be terminated with a comma to produce a valid
+# JSON. Hence, we add all but the last source in a foreach loop with comma as separate, while the
+# last item gets the empty string as separator.
+COMPILE_COMMANDS        :=
+ifeq (,$(SRCXX))
+  EXT                   := c
+  SRC_C_EXCEPT_LAST     := $(filter-out $(lastword $(SRC_C_ALL)),$(SRC_C_ALL))
+  ifneq (,$(SRC_C_EXCEPT_LAST))
+    COMPILE_COMMANDS    := $(foreach src,$(SRC_C_EXCEPT_LAST),$(call ADD_SEPERATOR,$(src),$(COMMA)))
+  endif
+  COMPILE_COMMANDS      += $(call ADD_SEPERATOR,$(lastword $(SRC_C_ALL)),)
+else
```

Comment:
I am a bit rusty on the details, is it not allowed to have both .c and .cpp files in the same module?

Reply:
It does. And generation of `compile_commands.json` works for mixed modules.

The if / else is only about for which source file to *not* add the comma after the compile command object. If there is only C code, I handle the last C source file separately. If there is C++ code, the last C++ file is added without the comma.